Slider on Android 5.0 breaks due to bad cast

We downcast the result of `getProgressDrawable` from `Drawable` to `LayerDrawable`. However, on Android 5.0 the returned result is a `StateListDrawable` (see stack trace below). This prevents compiled apps with non-Classic themes from working on Android 5.0 devices.

Stack trace:
```
E/AndroidRuntime( 3486): java.lang.RuntimeException: Unable to start activity ComponentInfo{appinventor.ai_ewpatton.SliderTest/appinventor.ai_ewpatton.SliderTest.Screen1}: java.lang.ClassCastException: android.graphics.drawable.StateListDrawable cannot be cast to android.graphics.drawable.LayerDrawable
E/AndroidRuntime( 3486): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2298)
E/AndroidRuntime( 3486): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2360)
E/AndroidRuntime( 3486): 	at android.app.ActivityThread.access$800(ActivityThread.java:144)
E/AndroidRuntime( 3486): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1278)
E/AndroidRuntime( 3486): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime( 3486): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime( 3486): 	at android.app.ActivityThread.main(ActivityThread.java:5221)
E/AndroidRuntime( 3486): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 3486): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime( 3486): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
E/AndroidRuntime( 3486): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
E/AndroidRuntime( 3486): Caused by: java.lang.ClassCastException: android.graphics.drawable.StateListDrawable cannot be cast to android.graphics.drawable.LayerDrawable
E/AndroidRuntime( 3486): 	at com.google.appinventor.components.runtime.Slider.<init>(Slider.java:99)
E/AndroidRuntime( 3486): 	at java.lang.reflect.Constructor.newInstance(Native Method)
E/AndroidRuntime( 3486): 	at java.lang.reflect.Constructor.newInstance(Constructor.java:288)
E/AndroidRuntime( 3486): 	at gnu.expr.PrimProcedure.apply(PrimProcedure.java:272)
E/AndroidRuntime( 3486): 	at gnu.mapping.CallContext.runUntilDone(CallContext.java:234)
E/AndroidRuntime( 3486): 	at gnu.mapping.CallContext.runUntilValue(CallContext.java:298)
E/AndroidRuntime( 3486): 	at gnu.kawa.reflect.Invoke.applyN(Invoke.java:209)
E/AndroidRuntime( 3486): 	at gnu.mapping.ProcedureN.apply2(ProcedureN.java:39)
E/AndroidRuntime( 3486): 	at appinventor.ai_ewpatton.SliderTest.Screen1.$define(Screen1.yail:10182)
E/AndroidRuntime( 3486): 	at com.google.appinventor.components.runtime.Form.onCreateFinish(Form.java:338)
E/AndroidRuntime( 3486): 	at com.google.appinventor.components.runtime.Form.onCreate(Form.java:285)
E/AndroidRuntime( 3486): 	at android.app.Activity.performCreate(Activity.java:5937)
E/AndroidRuntime( 3486): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1105)
E/AndroidRuntime( 3486): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2251)
E/AndroidRuntime( 3486): 	... 10 more
```

--- a/appinventor/components/src/com/google/appinventor/components/runtime/Slider.java
+++ b/appinventor/components/src/com/google/appinventor/components/runtime/Slider.java
@@ -1,9 +1,14 @@
 package com.google.appinventor.components.runtime;
 
 import android.R;
+import android.content.res.ColorStateList;
 import android.graphics.PorterDuff;
+import android.graphics.PorterDuff.Mode;
 import android.graphics.drawable.Drawable;
 import android.graphics.drawable.LayerDrawable;
+import android.graphics.drawable.StateListDrawable;
+import android.os.Build.VERSION;
+import android.os.Build.VERSION_CODES;
 import android.util.Log;
 import android.view.MotionEvent;
 import android.view.View;
@@ -57,11 +62,6 @@ public class Slider extends AndroidViewComponent implements SeekBar.OnSeekBarCha
   private float thumbPosition;
   private boolean thumbEnabled;
 
-  // the total slider width
-  private LayerDrawable fullBar;
-  // the part of the slider to the left of the thumb
-  private Drawable beforeThumb;
-
   // colors of the bar after and before the thumb position
   private int rightColor;
   private int leftColor;
@@ -100,9 +100,6 @@ public class Slider extends AndroidViewComponent implements SeekBar.OnSeekBarCha
     if (SdkLevel.getLevel() >= SdkLevel.LEVEL_LOLLIPOP) {
       seekbar.setSplitTrack(false);
     }
-
-    fullBar = (LayerDrawable) seekbar.getProgressDrawable();
-    beforeThumb = fullBar.findDrawableByLayerId(R.id.progress);
 
     leftColor = initialLeftColor;
     rightColor = initialRightColor;
@@ -143,9 +140,32 @@ public class Slider extends AndroidViewComponent implements SeekBar.OnSeekBarCha
   // NOTE(hal): On old phones, up through 2.2.2 and maybe higher, the color of the bar doesn't
   // change until the thumb is moved.  I'm ignoring that problem.
   private void setSliderColors() {
-   fullBar.setColorFilter(rightColor,PorterDuff.Mode.SRC);
-   beforeThumb.setColorFilter(leftColor, PorterDuff.Mode.SRC);
- }
+    if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP) {
+      seekbar.setProgressTintList(ColorStateList.valueOf(leftColor));
+      if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP_MR1 ||
+          !(seekbar.getProgressDrawable() instanceof StateListDrawable)) {
+        seekbar.setProgressBackgroundTintList(ColorStateList.valueOf(rightColor));
+        seekbar.setProgressBackgroundTintMode(Mode.MULTIPLY);
+      } else {
+        // Looking at the AOSP code, the previous calls should effectively accomplish what the
+        // following code does... except it doesn't on Android 5.0. Instead, the result is that the
+        // right side color is 50% opacity of leftColor. The following code works on Android 5.0,
+        // but assumes a Drawable hierarchy that may not be true if the device manufacturer deviates
+        // from the AOSP design. If that is the case, then the right hand side will not change.
+        StateListDrawable drawable = (StateListDrawable) seekbar.getProgressDrawable();
+        if (drawable.getCurrent() instanceof LayerDrawable) {
+          LayerDrawable layerDrawable = (LayerDrawable) drawable.getCurrent();
+          Drawable background = layerDrawable.findDrawableByLayerId(R.id.background);
+          background.setTintList(ColorStateList.valueOf(rightColor));
+          background.setTintMode(Mode.MULTIPLY);
+        }
+      }
+    } else {
+      LayerDrawable fullBar = (LayerDrawable) seekbar.getProgressDrawable();
+      fullBar.setColorFilter(rightColor,PorterDuff.Mode.SRC);
+      fullBar.findDrawableByLayerId(R.id.progress).setColorFilter(leftColor, PorterDuff.Mode.SRC);
+    }
+  }
 
  // Set the seekbar position based on minValue, maxValue, and thumbPosition
  // seekbar position is an integer in the range [0,100] and is determined by MinValue,

